coroutines-extra/src/main/scala/org/coroutines/extra/AsyncAwait.scala

```
@@ -0,0 +1,104 @@
+import scala.annotation.unchecked.uncheckedVariance
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent._
+import scala.language.experimental.macros
+import scala.reflect.macros.whitebox.Context
+import scala.util.{ Success, Failure }
+
+
+
+object AsyncAwait {
+  /** Await the result of a future.
+   *
+   *  When called inside an `async` body, this function will block until its
+   *  associated future completes.
+   *
+   *  @return A coroutine that yields a tuple. `async` will assign this tuple's
+   *          second element to hold the completed result of the `Future` passed
+   *          into the coroutine. The coroutine will directly return the
+   *          result of the future.
+   */
+  def await[R]: Future[R] ~~> (Future[R], R) =
+    coroutine { (awaitedFuture: Future[R]) =>
+      yieldval(awaitedFuture)
+      var result: R = null.asInstanceOf[R]
+      awaitedFuture.value match {
+        case Some(Success(x)) => result = x
+        case Some(Failure(error)) => throw error
+        case None => sys.error("Future was not completed")
+      }
+      result
+    }
+
+  /** Calls `body`, blocking on any calls to `await`.
+   *
+   *  @param body  A coroutine to be invoked.
+   *  @return      A `Future` wrapping the result of the coroutine. The future fails
+   *               if `body` throws an exception or one of the `await`s takes a failed
+   *               future.
+   */
+  def asyncCall[Y, R](body: ~~~>[Future[Y], R]): Future[R] = {
+    val c = call(body())
+    val p = Promise[R]
+    def loop() {
+      if (!c.resume) {
+        c.tryResult match {
+          case Success(result) => p.success(result)
+          case Failure(exception) => p.failure(exception)
+        }
+      } else {
+        val awaitedFuture = c.value
+        if (awaitedFuture.isCompleted) {
+          awaitedFuture.value match {
+            case Some(Success(result)) =>
+              loop()
+            case Some(Failure(exception)) =>
```

Why do you fail the promise at this point? What if a try-catch block wraps the `await` call? You should just resume and let the `await` throw the exception.

Can you please add a unit test for this (`await` call that should throw an exception, which is caught by the enclosing `try`)?